exampleCustomAudioDevice/src/main/java/com/twilio/examplecustomaudiodevice/FileAndMicAudioDevice.java

```diff
@@ -0,0 +1,372 @@
+        return isMusicPlaying;
+    }
+
+    /*
+     * Return the AudioFormat used the capturer. This custom device uses 44.1kHz sample rate and
+     * STEREO channel configuration both for microphone and the music file.
+     */
+    @Nullable
+    @Override
+    public AudioFormat getCapturerFormat() {
+        return new AudioFormat(AudioFormat.AUDIO_SAMPLE_RATE_44100,
+                AudioFormat.AUDIO_SAMPLE_STEREO);
+    }
+
+    /*
+     * Init the capturer using the AudioFormat return by getCapturerFormat().
+     */
+    @Override
+    public boolean onInitCapturer() {
+        int bytesPerFrame = 2 * (BITS_PER_SAMPLE / 8);
+        int framesPerBuffer = getCapturerFormat().getSampleRate() / BUFFERS_PER_SECOND;
+        // Calculate the minimum buffer size required for the successful creation of
+        // an AudioRecord object, in byte units.
+        int channelConfig = channelCountToConfiguration(getCapturerFormat().getChannelCount());
+        int minBufferSize =
+                AudioRecord.getMinBufferSize(getCapturerFormat().getSampleRate(),
+                        channelConfig, android.media.AudioFormat.ENCODING_PCM_16BIT);
+        micWriteBuffer = ByteBuffer.allocateDirect(bytesPerFrame * framesPerBuffer);
+        int bufferSizeInBytes = Math.max(BUFFER_SIZE_FACTOR * minBufferSize, micWriteBuffer.capacity());
+        audioRecord = new AudioRecord(MediaRecorder.AudioSource.MIC, getCapturerFormat().getSampleRate(),
+                android.media.AudioFormat.CHANNEL_OUT_STEREO, android.media.AudioFormat.ENCODING_PCM_16BIT, bufferSizeInBytes);
+
+        fileWriteByteBuffer = ByteBuffer.allocateDirect(bytesPerFrame * framesPerBuffer);
+        writeBufferSize = fileWriteByteBuffer.capacity();
+        // Initialize the streams.
+        initializeStreams();
+        return true;
+    }
+
+    @Override
+    public boolean onStartCapturing(@NonNull AudioDeviceContext audioDeviceContext) {
+        // Initialize the AudioDeviceContext
+        this.capturingAudioDeviceContext = audioDeviceContext;
+        // Create the capturer thread and start
+        capturerThread = new HandlerThread("CapturerThread");
+        capturerThread.start();
+        // Create the capturer handler that processes the capturer Runnables.
+        capturerHandler = new Handler(capturerThread.getLooper());
+        capturerHandler.post(fileCapturerRunnable);
+        return true;
+    }
+
+    @Override
+    public boolean onStopCapturing() {
+        if (isMusicPlaying) {
+            closeStreams();
+        } else {
+            stopRecording();
+        }
+        // Quit the capturerThread's looper to stop processing any further messages.
+        capturerThread.quit();
+        // Wait for the capturerThread to die.
+        if (!tvo.webrtc.ThreadUtils.joinUninterruptibly(capturerThread, THREAD_JOIN_TIMEOUT_MS)) {
+            Log.e(TAG, "Join of capturerThread timed out");
+            return false;
+        }
```

Comment:
Excellent improvement, but let's try a little less morbid language 😅 

```suggestion
        /*
         * When onStopCapturing is called, the AudioDevice API expects that at the completion
         * of the callback the capturer has completely stopped. As a result, quit the capturer
         * thread and explicitly wait for the thread to complete.
         */
        capturerThread.quit();
        if (!tvo.webrtc.ThreadUtils.joinUninterruptibly(capturerThread, THREAD_JOIN_TIMEOUT_MS)) {
            Log.e(TAG, "Join of capturerThread timed out");
            return false;
        }
```